### PR TITLE
[Messenger] Allow retries to be logged as warnings

### PR DESCRIPTION
**Pull request by jorissteyn**
This commit introduces a marker interface for exceptions to allow retries to be logged as warnings instead of the default error severity.

This behaviour is already implemented for recoverable exceptions, but those are retried ad infinitum regardless of the retry strategy. With the new marker interface, it's possible to log exceptions as warnings while still using the retry strategy.

The RecoverableExceptionInterface now extends this interface to achieve the existing behaviour using the new more flexible approach.

```
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Related to problem described in comments of https://github.com/symfony/messenger/pull/15
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/15844
```

### CODE REVIEW

**Comment by okwinza** [comment]
Hi @jorissteyn and thanks for your PR! 

One question: is there a reason you can't just throw `RecoverableExceptionInterface` to achieve the desired behavior?

File: src/Symfony/Component/Messenger/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 5.4
 ---
 
+ * Add the `LogRetryAsWarningInterface` for exceptions to log retries as warnings instead of errors allowing the same behaviour as RecoverableExceptionInterface while still using the retry strategy
```
Comment (okwinza) [inline src/Symfony/Component/Messenger/CHANGELOG.md]: Better move this to the bottom of the list.

File: src/Symfony/Component/Messenger/Exception/LogRetryAsWarningInterface.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+namespace Symfony\Component\Messenger\Exception;
+
+/**
+ * Marker interface for exceptions to indicate a retry should be logged as warning instead of error.
+ *
+ * @author Joris Steyn <j.steyn@hoffelijk.nl>
+ */
+interface LogRetryAsWarningInterface
```
Comment (okwinza) [inline src/Symfony/Component/Messenger/Exception/LogRetryAsWarningInterface.php]:
```suggestion
interface LogRetryAsWarningInterface extends \Throwable
```

File: src/Symfony/Component/Messenger/Exception/RecoverableExceptionInterface.php [hunk 3]
```diff
@@ -19,6 +19,6 @@
  *
  * @author Jérémy Derussé <jeremy@derusse.com>
  */
-interface RecoverableExceptionInterface extends \Throwable
+interface RecoverableExceptionInterface extends LogRetryAsWarningInterface, \Throwable
```
Comment (okwinza) [inline src/Symfony/Component/Messenger/Exception/RecoverableExceptionInterface.php]:
```suggestion
interface RecoverableExceptionInterface extends LogRetryAsWarningInterface
```

File: src/Symfony/Component/Messenger/Tests/EventListener/SendFailedMessageForRetryListenerTest.php [hunk 4]
```diff
@@ -108,10 +111,67 @@ public function testEnvelopeIsSentToTransportOnRetry()
         $retryStrategyLocator->expects($this->once())->method('has')->willReturn(true);
         $retryStrategyLocator->expects($this->once())->method('get')->willReturn($retryStategy);
 
+        $logger = $this->createMock(LoggerInterface::class);
```
Comment (okwinza) [inline src/Symfony/Component/Messenger/Tests/EventListener/SendFailedMessageForRetryListenerTest.php:114]: This seems to be out of scope of this particular test case.

**Review by okwinza — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Messenger/CHANGELOG.md (hunk 1 above)
Comment (jorissteyn) [reply src/Symfony/Component/Messenger/CHANGELOG.md]: done

File: src/Symfony/Component/Messenger/Exception/RecoverableExceptionInterface.php (hunk 3 above)
Comment (jorissteyn) [reply src/Symfony/Component/Messenger/Exception/RecoverableExceptionInterface.php]: Good one!

File: src/Symfony/Component/Messenger/Tests/EventListener/SendFailedMessageForRetryListenerTest.php (hunk 4 above)
Comment (jorissteyn) [reply src/Symfony/Component/Messenger/Tests/EventListener/SendFailedMessageForRetryListenerTest.php:114]: Right, restored this test to original version.

**Comment by jorissteyn** [comment]
> Hi @jorissteyn and thanks for your PR!
> 
> One question: is there a reason you can't just throw `RecoverableExceptionInterface` to achieve the desired behavior?

Yes, our goal is to log warnings for messages which are eventually successfully handled. Currently messenger allows for three use-cases:

 - unrecoverable, logged as critical
 - recoverable, always logged as warning, but might never be successfully handled
 - other exceptions: logged as error, but might succeed on retry

The new interface allows creating "retryable" exceptions, logged as warnings (like recoverable) until they fail on their last retry.

I've considered simply proposing "other exceptions" to be logged as warnings until the last retry, but that would be a BC break, and for "unexpected" exceptions, it makes sense to log them as errors even if they are retried. So expclicitly marking an exception as "log as warning" fills the gap. Messenger could also provide a "RetryableException" to be more in line with how recoverable works.

Does this make sense?

**Comment by jorissteyn** [comment]
After rebase on latest 5.4, the ` PHPUnit / Tests (7.2)` task is failing. I can't correlate the error with a change in this PR, also, runs fine on my machine :)

**Comment by nicolas-grekas** [comment]
Can you please share an example where this would be useful?
I'm not convinced by the naming, which looks too coupled to a specific use case.
What about adding a new `getLogLevel(\Exception $e)` method to `RetryStrategyInterface` instead of a marker interface?

**Comment by fabpot** [comment]
@jorissteyn Any comments on Nicolas's feedback?

**Comment by jorissteyn** [comment]
@fabpot I was working under the assumption the default behaviour (logging as error by default) was desirable to leave unchanged. The patch merged in https://github.com/symfony/symfony/pull/43492 (simply changing the default) makes a lot of sense and obsoletes the changes proposed in this thread.

I have however encountered projects that require more control in how different domain-specific exceptions are logged and @nicolas-grekas suggestion would accomplish that. I've taken a stab at it in https://github.com/symfony/symfony/pull/44120 and believe it's actually a nice and flexible solution while at the same time making the default behaviour much clearer code-wise.

> Can you please share an example where this would be useful?

Two abstract examples I listed in the new PR are:

 - Having specific exceptions logged as EMERGENCY
 - Log exceptions as NOTICE until or including the last attempt

Please don't hesitate to discuss this further if you're not fully convinced, I could elaborate on concrete use-cases I believe would be possible if we go forward with https://github.com/symfony/symfony/pull/44120.
